pkg/scheduler/algorithm/predicates/csi_volume_predicate.go

```diff
@@ -51,55 +54,52 @@ func NewCSIMaxVolumeLimitPredicate(
 func (c *CSIMaxVolumeLimitChecker) attachableLimitPredicate(
 	pod *v1.Pod, meta PredicateMetadata, nodeInfo *schedulernodeinfo.NodeInfo) (bool, []PredicateFailureReason, error) {
 
-	// if feature gate is disable we return
 	if !utilfeature.DefaultFeatureGate.Enabled(features.AttachVolumeLimit) {
 		return true, nil, nil
 	}
+
 	// If a pod doesn't have any volume attached to it, the predicate will always be true.
-	// Thus we make a fast path for it, to avoid unnecessary computations in this case.
+	// Thus we make a fast path for it to avoid unnecessary computations in this case.
 	if len(pod.Spec.Volumes) == 0 {
 		return true, nil, nil
 	}
 
-	nodeVolumeLimits := nodeInfo.VolumeLimits()
-
-	// if node does not have volume limits this predicate should exit
-	if len(nodeVolumeLimits) == 0 {
-		return true, nil, nil
-	}
-
-	// a map of unique volume name/csi volume handle and volume limit key
+	// A map of unique volume name/csi volume handle and volume limit key.
 	newVolumes := make(map[string]string)
-	if err := c.filterAttachableVolumes(pod.Spec.Volumes, pod.Namespace, newVolumes); err != nil {
+	if err := c.filterAttachableVolumes(nodeInfo, pod.Spec.Volumes, pod.Namespace, newVolumes); err != nil {
 		return false, nil, err
 	}
 
 	if len(newVolumes) == 0 {
 		return true, nil, nil
 	}
 
-	// a map of unique volume name/csi volume handle and volume limit key
+	// A map of unique volume name/csi volume handle and volume limit key.
 	attachedVolumes := make(map[string]string)
 	for _, existingPod := range nodeInfo.Pods() {
-		if err := c.filterAttachableVolumes(existingPod.Spec.Volumes, existingPod.Namespace, attachedVolumes); err != nil {
+		if err := c.filterAttachableVolumes(nodeInfo, existingPod.Spec.Volumes, existingPod.Namespace, attachedVolumes); err != nil {
 			return false, nil, err
 		}
 	}
 
-	newVolumeCount := map[string]int{}
 	attachedVolumeCount := map[string]int{}
-
-	for volumeName, volumeLimitKey := range attachedVolumes {
-		if _, ok := newVolumes[volumeName]; ok {
-			delete(newVolumes, volumeName)
+	for volumeUniqueName, volumeLimitKey := range attachedVolumes {
+		if _, ok := newVolumes[volumeUniqueName]; ok {
+			delete(newVolumes, volumeUniqueName)
 		}
 		attachedVolumeCount[volumeLimitKey]++
 	}
 
+	newVolumeCount := map[string]int{}
 	for _, volumeLimitKey := range newVolumes {
 		newVolumeCount[volumeLimitKey]++
 	}
 
+	nodeVolumeLimits := nodeInfo.VolumeLimits()
+	if len(nodeVolumeLimits) == 0 {
+		return true, nil, nil
+	}
+
```

Comment:
Now that we don't care about preventing pod scheduling for volumes where driver is not installed, it might be cheaper to move this *after* `newVolumes == 0` check and return early. Evaluating all pods on a node is def. going to be costlier.

Reply:
Moved up